### PR TITLE
Fix UAT tests re contracts

### PR DESCRIPTION
**Pull request by lucafavatella**
In scope for https://www.pivotaltracker.com/story/show/155510659

### CODE REVIEW

**Comment by hanssv** [comment]
Yes, it should be `aect_create_tx`...

**Comment by lucafavatella** [comment]
Yea still WIP

**Comment by lucafavatella** [comment]
As of 30be674 I see the following failing consistently and reproduced locally:
```
======================================================================
FAIL: integration.test_unsigned_tx.test_contract_create
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/builder/epoch/py/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/builder/epoch/py/tests/integration/test_unsigned_tx.py", line 63, in test_contract_create
    assert_equals(alice_balance0, alice_balance + test_settings["create_contract"]["fee"])
AssertionError: 20 != 31

======================================================================
FAIL: integration.test_unsigned_tx.test_contract_call
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/builder/epoch/py/local/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/builder/epoch/py/tests/integration/test_unsigned_tx.py", line 92, in test_contract_call
    assert_equals(alice_balance0, alice_balance + create_settings["create_contract"]["fee"])
AssertionError: 70 != 81
```

File: py/tests/integration/test_contracts.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def test_compile_id():
     contract = Contract( contract_string, "")
     result = api.compile_contract(contract)
 
-    bytecode = '0x36600080376200002160008080805180516004146200003057505b5060011951005b60005260206000f35b80905090565b602001517f6d61696e000000000000000000000000000000000000000000000000000000001462000061576200001a565b602001519050809150506200002a56'
+    bytecode = '0x366000803762000020600080805180516004146200002e57505b5060011951005b805190602001f35b80905090565b602001517f6d61696e00000000000000000000000000000000000000000000000000000000146200005f5762000019565b60200151806200006f9062000028565b5960200160008152818162000085918091505090565b81528059036020820390815292505050905090565b825180599081525060208401602084038393509350935050600082136200009a5780925050509056'
```
Comment (lucafavatella) [inline py/tests/integration/test_contracts.py]: I see difference can be traced to PR #797. The first step that differs is the icode, that carries type information. I also see [significant changes in aer_icode_to_asm.erl](https://github.com/aeternity/epoch/pull/797/files#diff-7c6a4e5f2d872793f9ddfe8a03886d33) so change in bytecode is justified. I am considering whether to make this test more robust to compiler changes (in a subsequent iteration).

----

Before PR #797: 30be674b (based on master f7aa14b5 with some UAT test changes)
After PR #797: d43547cf (master. Actually with a small UAT test change re `aect_create_tx`)


Same pp_ring_code:
```
contract type Identity = {
  type state;

  let main: int => int;
};

contract Identity = {
  let main(x:int) = {
    x;
  };
};
```

Same pp_ast:
```
[{contract_type,[{line,2}],
                {con,[{line,2}],"Identity"},
                [{type_decl,[{line,3}],{id,[{line,3}],"state"},[]},
                 {fun_decl,[{line,5}],
                           {id,[{line,5}],"main"},
                           {fun_t,[{line,5}],
                                  [{id,[{line,5}],"int"}],
                                  {id,[{line,5}],"int"}}}]},
 {contract,[{line,8}],
           {con,[{line,8}],"Identity"},
           [{letfun,[{line,9}],
                    {id,[{line,9}],"main"},
                    [{arg,[{line,9}],
                          {id,[{line,9}],"x"},
                          {id,[{line,9}],"int"}}],
                    {id,[{origin,system}],"_"},
                    {id,[{line,10}],"x"}}]}]
```

Same `aer_ast_infer_types` output:
```
Inferred "main" : (int) => int
```

Different pp_icode. Notice `word` at the end, from [this change](https://github.com/aeternity/epoch/commit/723d07462de3fc5ac7cab4374b4b96c4b5ce5f7d#diff-af0faefa2c5fbd841143fb60c7d1ced1R133):
* Before PR #797:
```
#{contract_name => "Identity",env => [],
  functions => [{"main",[{"x",{id,[{line,9}],"int"}}],{var_ref,"x"}}],
  options => [pp_bytecode,pp_assembler,pp_icode,pp_ast,pp_ring_code]}
```
* After PR #797:
```
#{contract_name => "Identity",env => [],
  functions => [{"main",[{"x",{id,[{line,9}],"int"}}],{var_ref,"x"},word}],
  options => [pp_bytecode,pp_assembler,pp_icode,pp_ast,pp_ring_code]}
```

In aer_icode_to_asm module:
* Only after PR #797:
```
Compiling {funcall,{var_ref,"_copymem"},
                   [{binop,'+',{var_ref,"addr"},{integer,32}},
                    {binop,'-',{var_ref,"length"},{integer,32}},
                    {var_ref,"result"}]}
 in stack [{"word","_"},{"result","_"},{"length","_"},{"addr","_"}]
       to ['PUSH1',32,'DUP5','ADD','PUSH1',32,'DUP5','SUB','DUP4','SWAP4',
           'POP','SWAP4','POP','SWAP4','POP','POP',
           {push_label,#Ref<0.1310447560.690749460.71355>},
           'JUMP']
```

Different pp_assembler:
* Before PR #797:
```
        CALLDATASIZE        PUSH1        0
        DUP1        CALLDATACOPY        PUSH3        0, 0, 33
        PUSH1        0
        DUP1        DUP1        DUP1        MLOAD        DUP1        MLOAD        PUSH1        4
        EQ        PUSH3        0, 0, 48
        JUMPI        POP        JUMPDEST        POP        PUSH1        1
        NOT        MLOAD        STOP        JUMPDEST        PUSH1        0
        MSTORE        PUSH1        32
        PUSH1        0
        RETURN        JUMPDEST        DUP1        SWAP1        POP        SWAP1        JUMP        JUMPDEST        PUSH1        32
        ADD        MLOAD        PUSH32        109, 97, 105, 110, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
        EQ        PUSH3        0, 0, 97
        JUMPI        PUSH3        0, 0, 26
        JUMP        JUMPDEST        PUSH1        32
        ADD        MLOAD        SWAP1        POP        DUP1        SWAP2        POP        POP        PUSH3        0, 0, 42
        JUMP
```
* After PR #797:
```
        CALLDATASIZE        PUSH1        0
        DUP1        CALLDATACOPY        PUSH3        0, 0, 32
        PUSH1        0
        DUP1        DUP1        MLOAD        DUP1        MLOAD        PUSH1        4
        EQ        PUSH3        0, 0, 46
        JUMPI        POP        JUMPDEST        POP        PUSH1        1
        NOT        MLOAD        STOP        JUMPDEST        DUP1        MLOAD        SWAP1        PUSH1        32
        ADD        RETURN        JUMPDEST        DUP1        SWAP1        POP        SWAP1        JUMP        JUMPDEST        PUSH1        32
        ADD        MLOAD        PUSH32        109, 97, 105, 110, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
        EQ        PUSH3        0, 0, 95
        JUMPI        PUSH3        0, 0, 25
        JUMP        JUMPDEST        PUSH1        32
        ADD        MLOAD        DUP1        PUSH3        0, 0, 111
        SWAP1        PUSH3        0, 0, 40
        JUMP        JUMPDEST        MSIZE        PUSH1        32
        ADD        PUSH1        0
        DUP2        MSTORE        DUP2        DUP2        PUSH3        0, 0, 133
        SWAP2        DUP1        SWAP2        POP        POP        SWAP1        JUMP        JUMPDEST        DUP2        MSTORE        DUP1        MSIZE        SUB        PUSH1        32
        DUP3        SUB        SWAP1        DUP2        MSTORE        SWAP3        POP        POP        POP        SWAP1        POP        SWAP1        JUMP        JUMPDEST        DUP3        MLOAD        DUP1        MSIZE        SWAP1        DUP2        MSTORE        POP        PUSH1        32
        DUP5        ADD        PUSH1        32
        DUP5        SUB        DUP4        SWAP4        POP        SWAP4        POP        SWAP4        POP        POP        PUSH1        0
        DUP3        SGT        PUSH3        0, 0, 154
        JUMPI        DUP1        SWAP3        POP        POP        POP        SWAP1        JUMP
```

Different pp_bytecode:
* Before PR #797:
```
0x00000000   CALLDATASIZE     
0x00000001   PUSH1            0x00
0x00000003   DUP1             
0x00000004   CALLDATACOPY     
0x00000005   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000021
0x00000009   PUSH1            0x00
0x0000000B   DUP1             
0x0000000C   DUP1             
0x0000000D   DUP1             
0x0000000E   MLOAD            
0x0000000F   DUP1             
0x00000010   MLOAD            
0x00000011   PUSH1            0x04
0x00000013   EQ               
0x00000014   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000030
0x00000018   JUMPI            
0x00000019   POP              
0x0000001A   JUMPDEST         
0x0000001B   POP              
0x0000001C   PUSH1            0x01
0x0000001E   NOT              
0x0000001F   MLOAD            
0x00000020   STOP             
0x00000021   JUMPDEST         
0x00000022   PUSH1            0x00
0x00000024   MSTORE           
0x00000025   PUSH1            0x20
0x00000027   PUSH1            0x00
0x00000029   RETURN           
0x0000002A   JUMPDEST         
0x0000002B   DUP1             
0x0000002C   SWAP1             
0x0000002D   POP              
0x0000002E   SWAP1            
0x0000002F   JUMP             
0x00000030   JUMPDEST         
0x00000031   PUSH1            0x20
0x00000033   ADD              
0x00000034   MLOAD            
0x00000035   PUSH32           0x6D61696E00000000000000000000000000000000000000000000000000000000
0x00000056   EQ               
0x00000057   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000061
0x0000005B   JUMPI            
0x0000005C   PUSH3            0x000000000000000000000000000000000000000000000000000000000000001A
0x00000060   JUMP             
0x00000061   JUMPDEST         
0x00000062   PUSH1            0x20
0x00000064   ADD              
0x00000065   MLOAD            
0x00000066   SWAP1            
0x00000067   POP              
0x00000068   DUP1             
0x00000069   SWAP2            
0x0000006A   POP              
0x0000006B   POP              
0x0000006C   PUSH3            0x000000000000000000000000000000000000000000000000000000000000002A
0x00000070   JUMP             
```
* After PR #797:
```
0x00000000   CALLDATASIZE     
0x00000001   PUSH1            0x00
0x00000003   DUP1             
0x00000004   CALLDATACOPY     
0x00000005   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000020
0x00000009   PUSH1            0x00
0x0000000B   DUP1             
0x0000000C   DUP1             
0x0000000D   MLOAD            
0x0000000E   DUP1             
0x0000000F   MLOAD            
0x00000010   PUSH1            0x04
0x00000012   EQ               
0x00000013   PUSH3            0x000000000000000000000000000000000000000000000000000000000000002E
0x00000017   JUMPI            
0x00000018   POP              
0x00000019   JUMPDEST         
0x0000001A   POP              
0x0000001B   PUSH1            0x01
0x0000001D   NOT              
0x0000001E   MLOAD            
0x0000001F   STOP             
0x00000020   JUMPDEST         
0x00000021   DUP1             
0x00000022   MLOAD            
0x00000023   SWAP1            
0x00000024   PUSH1            0x20
0x00000026   ADD              
0x00000027   RETURN           
0x00000028   JUMPDEST         
0x00000029   DUP1             
0x0000002A   SWAP1            
0x0000002B   POP              
0x0000002C   SWAP1            
0x0000002D   JUMP             
0x0000002E   JUMPDEST         
0x0000002F   PUSH1            0x20
0x00000031   ADD              
0x00000032   MLOAD            
0x00000033   PUSH32           0x6D61696E00000000000000000000000000000000000000000000000000000000
0x00000054   EQ               
0x00000055   PUSH3            0x000000000000000000000000000000000000000000000000000000000000005F
0x00000059   JUMPI            
0x0000005A   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000019
0x0000005E   JUMP             
0x0000005F   JUMPDEST         
0x00000060   PUSH1            0x20
0x00000062   ADD              
0x00000063   MLOAD            
0x00000064   DUP1             
0x00000065   PUSH3            0x000000000000000000000000000000000000000000000000000000000000006F
0x00000069   SWAP1            
0x0000006A   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000028 
0x0000006E   JUMP             
0x0000006F   JUMPDEST         
0x00000070   MSIZE            
0x00000071   PUSH1            0x20
0x00000073   ADD              
0x00000074   PUSH1            0x00
0x00000076   DUP2             
0x00000077   MSTORE           
0x00000078   DUP2             
0x00000079   DUP2             
0x0000007A   PUSH3            0x0000000000000000000000000000000000000000000000000000000000000085
0x0000007E   SWAP2            
0x0000007F   DUP1             
0x00000080   SWAP2            
0x00000081   POP              
0x00000082   POP              
0x00000083   SWAP1            
0x00000084   JUMP             
0x00000085   JUMPDEST         
0x00000086   DUP2             
0x00000087   MSTORE           
0x00000088   DUP1             
0x00000089   MSIZE            
0x0000008A   SUB              
0x0000008B   PUSH1            0x20
0x0000008D   DUP3             
0x0000008E   SUB              
0x0000008F   SWAP1            
0x00000090   DUP2             
0x00000091   MSTORE           
0x00000092   SWAP3            
0x00000093   POP              
0x00000094   POP              
0x00000095   POP              
0x00000096   SWAP1            
0x00000097   POP              
0x00000098   SWAP1            
0x00000099   JUMP             
0x0000009A   JUMPDEST         
0x0000009B   DUP3             
0x0000009C   MLOAD            
0x0000009D   DUP1             
0x0000009E   MSIZE            
0x0000009F   SWAP1            
0x000000A0   DUP2             
0x000000A1   MSTORE           
0x000000A2   POP              
0x000000A3   PUSH1            0x20
0x000000A5   DUP5             
0x000000A6   ADD              
0x000000A7   PUSH1            0x20
0x000000A9   DUP5             
0x000000AA   SUB              
0x000000AB   DUP4             
0x000000AC   SWAP4            
0x000000AD   POP              
0x000000AE   SWAP4            
0x000000AF   POP              
0x000000B0   SWAP4            
0x000000B1   POP              
0x000000B2   POP              
0x000000B3   PUSH1            0x00
0x000000B5   DUP3             
0x000000B6   SGT              
0x000000B7   PUSH3            0x000000000000000000000000000000000000000000000000000000000000009A
0x000000BB   JUMPI            
0x000000BC   DUP1             
0x000000BD   SWAP3            
0x000000BE   POP              
0x000000BF   POP              
0x000000C0   POP              
0x000000C1   SWAP1            
0x000000C2   JUMP             
```

**Pull request merged**